### ISSUE TITLE
Great library, but ??!!

### ISSUE THREAD

**Issue by udayrddy** [camelot-dev/camelot#49]
_Stats below are pulled from PyPI downloads._ 
Despite being a better process than the others, what do you think supports the less usage.

![image](https://user-images.githubusercontent.com/33936764/62944891-a1049480-bdab-11e9-90a3-810d2904e889.png)

**Comment by vinayak-mehta** [comment]
Yep, this is a known issue. We need to figure out a way to replace ghostscript and opencv. https://github.com/camelot-dev/camelot/issues/13

Camelot uses only a small subset of code from ghostscript [1] (converting PDF to PNG) and opencv [2] (adaptive thresholding and morphological transformations). The only way I can think of is to re-implement these in Python and have them inside Camelot itself. [2] should be straightforward. Do you have any ideas around [1]? Or any other pointers?

Ghostscript is written in C, I tried looking around in the huge codebase but was totally lost. I'm planning to look into this by allotting time over the next month, currently the day job takes up a lot of the time. Any pointers would really help!

**Comment by vinayak-mehta** [comment]
@jnothman Do you have any pointers around this?

**Comment by sathk882** [comment]
@vinayak-mehta , Have you tried pdftoppm( poppler utils) for converting pdf to png.

**Comment by vinayak-mehta** [comment]
Yep, I tried it along with imagemagick before landing on ghostscript since the last one gave the best results in terms of image quality.

**Comment by sathk882** [comment]
Ok, In this [post](https://serverfault.com/questions/167573/fast-pdf-to-jpg-conversion-on-linux-wantedl), there was one more suggestion to do this with [MuPDF](https://mupdf.com/index.html)

**Comment by jnothman** [comment]
Hey @vinayak-mehta, not sure where I can help here!  The change to pdfbox in #30 has been implemented too, but you need to confirm what kinds of discrepancies are acceptable between backends.

**Comment by dimitern** [comment]
@akshowhini Camelot already is open source and MIT licensed. It looks like your CamelotPro uses some of the sources from Camelot (I guess to make it compatible), and is GPL 3.0 licensed. It will be nice to mention the original authors somewhere as well.

The SaaS backend you're using is proprietary licensed, but I supposed also uses Camelot.

I'd be interested to know more about the CamelotPro flavour, for example - does it autodetect whether it should use lattice or stream ?

**Comment by vinayak-mehta** [comment]
> If you think the service as an add-on helps the regular camelot users like me, I would be happy to talk with my team to merge CamelotPro with the open sourced lib

I think it would make sense to include CamelotPro in Camelot if former's code is open-sourced. (assuming it works well on the current tests and more image-based tests)

> I'd be interested to know more about the CamelotPro flavour, for example - does it autodetect whether it should use lattice or stream ?

I'd be interested in learning about the internal workings too!

I tried running the example you've provided in the README but it fails with a KeyError: https://github.com/ExtractTable/camelotpro/issues/1

**Comment by luke4u** [comment]
> converting PDF to PNG

Hi @vinayak-mehta , have you heard about pdf2image? It is much less pain than Ghostscript. I don't recommend mupdf, as it is licensed under AGPL. Not good for commercial usage.

**Comment by vinayak-mehta** [comment]
@luke4u Yes I've heard about it. I was initally reluctant to replace ghostscript with pdf2image as users on all platforms would have to install it separately too. Doing this replacement would be difficult as it would break installations for old users when they suddenly need poppler-utils instead of ghostscript after they upgrade their camelot version. I guess it could be done in a backwards-compatible manner where ghostscript and pdf2image are different backends that camelot can use based on availability.

I'll try to move fast on my goal of using a pdf -> png conversion tool as a depedency altogether and make the library self-contained.

**Comment by luke4u** [comment]
thank you @vinayak-mehta . look forward to your progress!

**Comment by HamedMP** [comment]
Trying to install `pdftopng` and getting this error. Worth using more stable dependencies:

```sh
  • Installing pdftopng (0.2.3): Failed

  RuntimeError

  Unable to find installation candidates for pdftopng (0.2.3)

  at ~/Library/Application Support/pypoetry/venv/lib/python3.11/site-packages/poetry/installation/chooser.py:73 in choose_for
       69│
       70│             links.append(link)
       71│
       72│         if not links:
    →  73│             raise RuntimeError(f"Unable to find installation candidates for {package}")
       74│
       75│         # Get the best link
       76│         chosen = max(links, key=lambda link: self._sort_key(package, link))
```

**Comment by agamm** [comment]
> Trying to install `pdftopng` and getting this error. Worth using more stable dependencies:
> 
> ```shell
>   • Installing pdftopng (0.2.3): Failed
> 
>   RuntimeError
> 
>   Unable to find installation candidates for pdftopng (0.2.3)
> 
>   at ~/Library/Application Support/pypoetry/venv/lib/python3.11/site-packages/poetry/installation/chooser.py:73 in choose_for
>        69│
>        70│             links.append(link)
>        71│
>        72│         if not links:
>     →  73│             raise RuntimeError(f"Unable to find installation candidates for {package}")
>        74│
>        75│         # Get the best link
>        76│         chosen = max(links, key=lambda link: self._sort_key(package, link))
> ```

I have the same error with `poetry add  "camelot-py[base]"`

**Comment by synweap15** [comment]
Same error here. Any resolution @HamedMP @agamm @sfrancis19?

**Comment by bosd** [comment]
Closing as resolved — the dependency concern that started this thread has been addressed. As of **v1.0.0** Camelot dropped the Ghostscript dependency and uses **pdfium** as the default image-conversion backend (poppler/Ghostscript are now optional extras), and it ships against **opencv-python-headless**. A plain `pip install camelot-py[base]` is now a light, wheels-only install on the common platforms — see https://camelot-py.readthedocs.io/en/latest/user/install.html . Thanks for kicking off the discussion that fed into #13.
